src/Components/Components/src/NavigationManagerExtensions.cs

```diff
@@ -0,0 +1,938 @@
+        /// <param name="values">The parameter values to add or update.</param>
+        /// <remarks>
+        /// Any <c>null</c> entries in <paramref name="values"/> will be skipped. Existing querystring parameters not in
+        /// <paramref name="values"/> will be removed from the querystring in the returned URI.
+        /// </remarks>
+        public static string UriWithQueryParameter(this NavigationManager navigationManager, string name, IEnumerable<long> values)
+            => UriWithQueryParameter(navigationManager, name, values, Format);
+
+        /// <summary>
+        /// Returns a URI that is constructed by updating <see cref="NavigationManager.Uri"/> with a single parameter
+        /// updated with the provided <paramref name="values"/>.
+        /// </summary>
+        /// <param name="navigationManager">The <see cref="NavigationManager"/>.</param>
+        /// <param name="name">The name of the parameter to add or update.</param>
+        /// <param name="values">The parameter values to add or update.</param>
+        /// <remarks>
+        /// Any <c>null</c> entries in <paramref name="values"/> will be skipped. Existing querystring parameters not in
+        /// <paramref name="values"/> will be removed from the querystring in the returned URI.
+        /// </remarks>
+        public static string UriWithQueryParameter(this NavigationManager navigationManager, string name, IEnumerable<long?> values)
+            => UriWithQueryParameter(navigationManager, name, values, Format);
+
+        /// <summary>
+        /// Returns a URI that is constructed by updating <see cref="NavigationManager.Uri"/> with a single parameter
+        /// added, updated, or removed.
+        /// </summary>
+        /// <param name="navigationManager">The <see cref="NavigationManager"/>.</param>
+        /// <param name="name">The name of the parameter to add or update.</param>
+        /// <param name="value">The value of the parameter to add or update.</param>
+        /// <remarks>
+        /// If <paramref name="value"/> is <c>null</c>, the parameter will be removed if it exists in the URI.
+        /// Otherwise, it will be added or updated.
+        /// </remarks>
+        public static string UriWithQueryParameter(this NavigationManager navigationManager, string name, string? value)
+        {
+            if (navigationManager is null)
+            {
+                throw new ArgumentNullException(nameof(navigationManager));
+            }
+
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new InvalidOperationException(EmptyQueryParameterNameExceptionMessage);
+            }
+
+            var uri = navigationManager.Uri;
+
+            return value is null
+                ? UriWithoutQueryParameter(uri, name)
+                : UriWithQueryParameterCore(uri, name, value);
+        }
+
+        private static string UriWithQueryParameter<TValue>(
```

Comment:
One slight drawback to these similarly-named internal methods (e.g., `UriWithQueryParameter`, `UriWithQueryParameterCore`, `UriWithAppendedQueryParameters`) is that you have to read the implementation in some detail to understand how they differ and relate to each other. If there were any more explicit names that were more clearly different it might help with skim-reading.

For example this could be something like `SetParameterWithMultipleValues`.

This is not at all a big deal and the PR would be fine to merge as-is. Not sure whether more explicit names are possible in other cases.

Reply:
Definitely - this is something I've thought about changing as well. I'll improve the naming of some of these internal methods to make things more clear.